osu.Game.Rulesets.Osu/OsuInputManager.cs

```diff
@@ -43,17 +53,52 @@ protected override bool Handle(UIEvent e)
             return base.Handle(e);
         }
 
+        private OsuAction getActionForTouchSource(TouchSource source)
+        {
+            int sourceNumber = numberedTouchSources[source];
+            return sourceNumber % 2 == 0 ? OsuAction.RightButton : OsuAction.LeftButton;
+        }
+
         protected override bool HandleMouseTouchStateChange(TouchStateChangeEvent e)
         {
-            if (!AllowUserCursorMovement)
+            var source = e.Touch.Source;
+
+            int touchNumber = numberedTouchSources[source];
+
+            if (touchNumber > simultaneous_touches_limit)
+                return false;
+
+            var activeSources = CurrentState.Touch.ActiveSources;
+
+            var cursorTouch = activeSources.Any() ? activeSources.First() : source;
+
+            var activeTapTouches = AllowUserCursorMovement ? activeSources.Skip(1) : activeSources;
+            var limitedActiveTapTouches = activeTapTouches.Take(tap_touches_limit);
+
+            var newActiveTapActions = limitedActiveTapTouches.Select(getActionForTouchSource);
+            var newInactiveTapActions = PressedActions.Where(a => !newActiveTapActions.Contains(a)).ToList();
+
+            bool isCursorTouch = source == cursorTouch;
+            bool isTapTouch = !isCursorTouch;
+
+            foreach (var action in newInactiveTapActions)
+                KeyBindingContainer.TriggerReleased(action);
+
+            if (isTapTouch)
+            {
+                var action = getActionForTouchSource(source);
```

Comment:
For consistency with this code, the action should be determined by the index of the touch in the active tap touches.
```suggestion
                int tapNumber = activeTapTouches.TakeWhile(t => t != source).Count();
                var action = tapNumber % 2 == 1 ? OsuAction.RightButton : OsuAction.LeftButton;

```